### PR TITLE
Segformer Semantic Segmentation

### PR DESCRIPTION
**Pull request by lucasreljic**
Node running:
The model is: "segformer_mit-b2_8xb1-160k_cityscapes-1024x1024"

https://github.com/WATonomous/wato_monorepo/assets/56940300/dca8fd7d-b54c-4bcd-9769-d9e548a11563

### CODE REVIEW

File: watod-config.sh
Comment (danielrhuynh) [inline watod-config.sh:1]: Remove this from PR

File: src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        # print(f'Publishing Segmentation')
+        # self.image_publisher.publish(img_output)
+        color_seg = cv2.resize(color_seg,(1600, 900))
+        mask_output = self.bridge.cv2_to_imgmsg(color_seg)
+        self.image_publisher.publish(mask_output)
+
+
+
+def main(args=None):
+
```
Comment (danielrhuynh) [inline src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py:136]: Fix autopep

File: src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+                    self.get_logger().error(str(e))
+                    return
+        
+        with torch.no_grad():
+            out_img = self.model(image, show=False, )['predictions']
+    
+        # logits = torch.tensor(out_img, dtype=torch.float32).unsqueeze(0).unsqueeze(0)  # Add batch and channel dimensions
+        # upsampled_logits = torch.nn.functional.interpolate(logits,
+        #                                                    size=(IMAGE_H, IMAGE_W),  # (height, width)
+        #                                                    mode='bilinear',
```
Comment (danielrhuynh) [inline src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py:115]: Fix autopep here

File: src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+IMAGE_H = 900
+IMAGE_W = 1600
+
+#Adjust logging
+logging.getLogger('mmseg.apis').setLevel(logging.WARNING)
+logging.getLogger('mmengine').setLevel(logging.WARNING)
+
+"""Cityscapes palette that maps each class to RGB values."""
+COLOR_PALLETE = [
```
Comment (danielrhuynh) [inline src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py:37]: Fix autopep here

File: src/perception/semantic_segmentation/resource/deploy.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Deploy:
```
Comment (danielrhuynh) [inline src/perception/semantic_segmentation/resource/deploy.yaml:1]: Remove?

File: src/perception/semantic_segmentation/resource/cityscapes_demo.png
Comment (danielrhuynh) [inline src/perception/semantic_segmentation/resource/cityscapes_demo.png:1]: Remove photo?

File: src/perception/semantic_segmentation/launch/semantic_segmentation.launch.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    semantic_segmentation_node = Node(
+            package='semantic_segmentation',
+            executable='semantic_segmentation_node',
+            name='semantic_segmentation_node',
+            parameters=[config]
+        )
+
+    # finalize
+    ld.add_action(semantic_segmentation_node)
+
```
Comment (danielrhuynh) [inline src/perception/semantic_segmentation/launch/semantic_segmentation.launch.py:24]: Fix autopep

File: docker/perception/semantic_segmentation/semantic_segmentation.Dockerfile [hunk 6]
```diff
@@ -43,11 +103,38 @@ RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
 
 # Entrypoint will run before any CMD on launch. Sources ~/opt/<ROS_DISTRO>/setup.bash and ~/ament_ws/install/setup.bash
 COPY docker/wato_ros_entrypoint.sh ${AMENT_WS}/wato_ros_entrypoint.sh
+
+WORKDIR /mmsegmentation
+COPY --from=Segformer /mmsegmentation /mmsegmentation
+RUN pip install -r requirements.txt
+RUN pip install --no-cache-dir -e .
+WORKDIR ${AMENT_WS}
+#  --from=Segformer /mmsegmentation/model /model
```
Comment (danielrhuynh) [inline docker/perception/semantic_segmentation/semantic_segmentation.Dockerfile:112]: clean up comments

File: docker/perception/semantic_segmentation/semantic_segmentation.Dockerfile [hunk 7]
```diff
@@ -1,13 +1,43 @@
...
+RUN apt-get update && apt-get install -y git ninja-build libglib2.0-0 libsm6 libxrender-dev libxext6 libgl1-mesa-dev  \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+WORKDIR ${AMENT_WS}/src
+# Install MMSegmentation
+RUN git clone -b main https://github.com/open-mmlab/mmsegmentation.git /mmsegmentation
+WORKDIR /mmsegmentation
+ENV FORCE_CUDA="1"
+#RUN pip install -r requirements.txt
```
Comment (danielrhuynh) [inline docker/perception/semantic_segmentation/semantic_segmentation.Dockerfile:21]: Clean comment

**Review by danielrhuynh — CHANGES_REQUESTED** [review]
Good stuff Lucas! Just a few catches from an overview before it gets to Eddy

File: src/perception/semantic_segmentation/resource/deploy.yaml (hunk 4 above)
Comment (lucasreljic) [reply src/perception/semantic_segmentation/resource/deploy.yaml]: Done

**Review by danielrhuynh — APPROVED** [review]
Nice work!

**Comment by danielrhuynh** [comment]
Wait for Eddie before merging tho

**Comment by Edwardius** [comment]
Since this is a transformer, how much vram does this take up?

File: src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+                history=QoSHistoryPolicy.KEEP_LAST,
+                depth=0,
+            ),
+        )
+
+        self.image_publisher = self.create_publisher(
+            Image,
+            '/camera/left/segmentations',
+            10
+        )
```
Comment (Edwardius) [inline src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py:88]: instead of outputting a segmented image, I would instead store the masks, in a custom message

File: src/samples/python/aggregator/setup.py [hunk 9]
```diff
@@ -14,7 +14,7 @@
         # Include our package.xml file
         (os.path.join('share', package_name), ['package.xml']),
         # Include all launch files.
-        (os.path.join('share', package_name, 'launch'), \
+        (os.path.join('share', package_name, 'launch'),
          glob(os.path.join('launch', '*.launch.py'))),
```
Comment (Edwardius) [inline src/samples/python/aggregator/setup.py:18]: unintentional?

File: src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+logging.basicConfig(level=logging.INFO, handlers=[RichHandler(level=logging.WARNING)])
+
+
+# ALGO_VERSION = os.getenv("MODEL_NAME")
+
+# if not ALGO_VERSION:
+#     ALGO_VERSION = 'nvidia/segformer-b2-finetuned-ade-512-512'
+
+package_name = 'semantic_segmentation'
+package_share_directory = get_package_share_directory(package_name)
```
Comment (Edwardius) [inline src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py:25]: this should happen in the launch file, same place as where the yaml config is loaded

File: src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    [220, 20, 60],   # person
+    [255, 0, 0],     # rider
+    [0, 0, 142],     # car
+    [0, 0, 70],      # truck
+    [0, 60, 100],    # bus
+    [0, 80, 100],    # train
+    [0, 0, 230],     # motorcycle
+    [119, 11, 32],   # bicycle
+    [0, 0, 0]        # ignore/unlabeled
+]
```
Comment (Edwardius) [inline src/perception/semantic_segmentation/semantic_segmentation/segmentation_node.py:59]: all of these should be in the config

**Review by Edwardius — CHANGES_REQUESTED** [review]
Great work! A large portion of my concerns have to do with the size of the docker image, as well as the ros messages being used. 

To reduce the size of the docker image, there are two avenues:
1. wrap your head around docker caching to make MMLab stuff work
2. use a simpler repo that has an implementation of segformer, ie. https://github.com/lucidrains/segformer-pytorch

File: src/samples/python/aggregator/setup.py (hunk 9 above)
Comment (danielrhuynh) [reply src/samples/python/aggregator/setup.py:18]: The linter was complaining about this?

**Comment by lucasreljic** [comment]
I moved all the config stuff in the yaml file, the docker file should be smaller now. The publisher message is still one Image but it's just the mask now. I can change to a boolean array (such as ByteMultiArray I think) or the new vision message you mentioned. Let me know which one I should implement.

**Review by Edwardius — APPROVED** [review]

**Comment by lucasreljic** [comment]
Should a LabelInfo topic still be created for semantics, to map the class name to the int value in the mask?
